src/openrct2-ui/windows/Viewport.cpp

```diff
@@ -10,192 +10,217 @@
+// clang-format on
 
-static rct_window_event_list window_viewport_events([](auto& events)
+class ViewportWindow final : public Window
 {
-    events.mouse_up = &window_viewport_mouseup;
-    events.resize = &window_viewport_resize;
-    events.update = &window_viewport_update;
-    events.invalidate = &window_viewport_invalidate;
-    events.paint = &window_viewport_paint;
-});
-// clang-format on
+private:
+    int32_t _viewportWindowNumber = 1;
 
-static int32_t _viewportNumber = 1;
+    void get_free_viewport_number(rct_window* w)
+    {
+        if (w != nullptr && w->classification == WC_VIEWPORT)
+        {
+            if (w->number >= _viewportWindowNumber)
+                _viewportWindowNumber = w->number + 1;
+        }
+    }
 
-/**
- * Creates a custom viewport window.
- */
-rct_window* window_viewport_open()
-{
-    rct_window* w = WindowCreateAutoPos(WW, WH, &window_viewport_events, WC_VIEWPORT, WF_RESIZABLE);
-    w->widgets = window_viewport_widgets;
-    w->enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
-    w->number = _viewportNumber++;
-
-    // Create viewport
-    viewport_create(w, w->windowPos, w->width, w->height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
-    rct_window* mainWindow = window_get_main();
-    if (mainWindow != nullptr)
+public:
+    void OnOpen() override
     {
-        rct_viewport* mainViewport = mainWindow->viewport;
-        int32_t x = mainViewport->viewPos.x + (mainViewport->view_width / 2);
-        int32_t y = mainViewport->viewPos.y + (mainViewport->view_height / 2);
-        w->savedViewPos = { x - (w->viewport->view_width / 2), y - (w->viewport->view_height / 2) };
+        window_visit_each([&](rct_window* w) { get_free_viewport_number(w); });
+        number = _viewportWindowNumber;
+
+        widgets = window_viewport_widgets;
+        enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
+
+        // Create viewport
+        viewport_create(this, windowPos, width, height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
+        rct_window* mainWindow = window_get_main();
+        if (mainWindow != nullptr)
+        {
+            rct_viewport* mainViewport = mainWindow->viewport;
+            int32_t x = mainViewport->viewPos.x + (mainViewport->view_width / 2);
+            int32_t y = mainViewport->viewPos.y + (mainViewport->view_height / 2);
+            savedViewPos = { x - (viewport->view_width / 2), y - (viewport->view_height / 2) };
+        }
+
+        viewport->flags |= VIEWPORT_FLAG_SOUND_ON;
+
+        min_width = WW;
+        min_height = WH;
+        max_width = WW;
+        max_height = WH;
     }
 
-    w->viewport->flags |= VIEWPORT_FLAG_SOUND_ON;
+    void OnUpdate() override
+    {
+        rct_window* mainWindow;
 
-    return w;
-}
+        mainWindow = window_get_main();
```

Comment:
As you improve this code, you can start reducing the number of unnecessary lines which were reminiscent of when the codebase was in C. So in this example you can just write `auto mainWindow = window_get_main();`

Reply:
Which is better / preferred?
1. `rct_window* mainWindow = window_get_main();`
or
2. `auto mainWindow = window_get_main();`

Reply:
I went with 2 `auto mainWindow = window_get_main();`